### ISSUE TITLE
Missing MPI collectives op symbols in TF build

### ISSUE THREAD

**Issue by jthestness** [tensorflow/tensorflow#13875]
Running into a build issue when trying to use MPI collectives. We are able to build a recent version of TF using NVIDIA's changes for CUDA 9/cuDNN 7. However, it appears that the build strips the MPI collectives library ops:

```
% python -c 'import tensorflow.contrib.mpi_collectives as mpi'
Traceback (most recent call last):
 File "<string>", line 1, in <module>
 File "/mnt/home/lib/python3.6/site-packages/tensorflow/contrib/mpi_collectives/__init__.py", line 126, in <module>
   from tensorflow.contrib.mpi_collectives.mpi_ops import size
 File "/mnt/home/lib/python3.6/site-packages/tensorflow/contrib/mpi_collectives/mpi_ops.py", line 59, in <module>
   'MPIAllreduce'])
 File "/mnt/home/lib/python3.6/site-packages/tensorflow/contrib/mpi_collectives/mpi_ops.py", line 51, in _load_library
   (expected_op, name))
NameError: Could not find operator MPISize in dynamic library mpi_collectives.so
```

It seems like the issue might be caused by commit 5c7f9e3, which changes linking behavior, but we're unable to bisect due to commit order dependencies.

Anyone have an idea how to fix the issue? We're willing to update the MPI collectives code and submit a PR fix.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: We've applied NVIDIA's CUDA 9/cuDNN 7 patches for mixed-precision per this page: http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#training_tensorflow
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 14.04
- **TensorFlow version (use command below)**: ea94bbe9fa9f9b3d01fb057c02ef7873d76bf09c
- **Python version**: 3.6
- **Bazel version**: 0.5.4
- **CUDA/cuDNN version**: CUDA 9.0.103_rc/cuDNN 7.0-rc
- **Exact command to reproduce**: `python -c 'import tensorflow.contrib.mpi_collectives as mpi'`

**Comment by allenlavoie** [comment]
One test/workaround is to build with `--config=monolithic`. That will give you the old op loading behavior.

Several questions come to mind, like why we're not running this code in tests regularly (are we?), and what the actual issue is with the op loading. I can take a look next week.

**Comment by jthestness** [comment]
We've tested a few more things: we tried `--config=monolithic`, we built the current TF mainline head ea94bbe without the CUDA 9/cuDNN 7 changes from NVIDIA, and built using prior versions of CUDA (8) and cuDNN (6). The error exists after each of these builds. IOW, the NVIDIA patches were not part of the problem.

We've started looking into Bazel's process for static linking against the MPI library. It seems like we might not be pulling MPI symbols into the TF binary, so during linking, gcc might be stripping out the MPI ops thinking they don't connect to anything (i.e., dead code elimination).

The reason we don't have any Jenkins tests for tf.contrib.mpi_collectives is that we don't currently have a way to build TF with MPI enabled. It requires an MPI library to run the simple ops, and to test the collectives ops, we would need Bazel to be able to run multiple threads or processes that would communicate.

**Comment by allenlavoie** [comment]
Does it work if you add alwayslink=1 to https://github.com/tensorflow/tensorflow/blob/53e7541cf7efa61ba22c9f042e07031d87c8f145/tensorflow/tensorflow.bzl#L1138 ?

**Comment by jthestness** [comment]
Just wanted to give an update since other have asked about this as well: We've tried a number of different combinations of `alwayslink`, `linkstatic`, and `linkshared`, but nothing has fixed this issue. I think we will need to restructure the Bazel targets to get finer-grained control of the binary or library options during build. I won't be able to look into this soon, but if anyone wants to use the MPI collectives in the meantime, we recommend building a TF mainline commit before 5c7f9e3

**Comment by allenlavoie** [comment]
So it does work right before that commit? I thought you couldn't bisect.

Can you try copying the "defines=" from //tensorflow/core:lib_internal to //tensorflow/core:lib_internal_impl (or just wait a bit because I have a commit which will sync and do that)? Defines were not applied to .cc files in that commit. That's the only difference I know of when building with --config=monolithic.

**Comment by jthestness** [comment]
After removing the CUDA 9/cuDNN 7 changes (by establishing they weren't the problem), we were able to bisect. We built with the commit just before 5c7f9e3 (i.e., 054b882), and the MPI collectives import works there.

Thanks for the recommendations! I'll give them a try sometime next week.

**Comment by NEWPLAN** [comment]
i roll back to commit 054b88233bf6d6bc5b953fca50dbb01d108b2d18, but there is not mpi_collectives in this branch...
i reset it wih command git reset --hard 054b88233bf6d6bc5b953fca50dbb01d108b2d18...
sorry for troubling you.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by jthestness** [comment]
I submitted a fix for this in PR #15534. Stay tuned there for updates.

**Comment by jthestness** [comment]
The fix #15534 is merged. `mpi_collectives` in mainline now works for us. Closing this issue.
